pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go

```diff
@@ -0,0 +1,465 @@
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	klog.Warningf("Retry budget exceeded, dropping %q Endpoints out of the queue: %v", key, err)
+	c.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+func (c *Controller) syncEndpoints(key string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing EndpointSlices for %q Endpoints. (%v)", key, time.Since(startTime))
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	esLabelSelector := labels.Set(map[string]string{
+		discovery.LabelServiceName: name,
+		discovery.LabelManagedBy:   controllerName,
+	}).AsSelectorPreValidated()
+	endpointSlices, err := c.endpointSliceLister.EndpointSlices(namespace).List(esLabelSelector)
+
+	if err != nil {
+		ep := &v1.Endpoints{ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: namespace}}
+		c.eventRecorder.Eventf(ep, "FailedToListEndpointSlices",
+			"Error listing EndpointSlices for Endpoints %s/%s: %v", ep.Namespace, ep.Name, err)
+		return err
+	}
+
+	endpoints, err := c.endpointsLister.Endpoints(namespace).Get(name)
+	if err != nil || skipMirror(endpoints.Labels) {
+		if apierrors.IsNotFound(err) || skipMirror(endpoints.Labels) {
+			return c.reconciler.deleteEndpoints(namespace, name, endpointSlices)
+		}
+		return err
+	}
+
+	err = c.reconciler.reconcile(endpoints, endpointSlices)
+	if err != nil {
+		c.eventRecorder.Eventf(endpoints, v1.EventTypeWarning, "FailedToUpdateEndpointSlices",
+			"Error updating EndpointSlices for Endpoints %s/%s: %v", endpoints.Namespace, endpoints.Name, err)
+		return err
+	}
+
+	return nil
+}
+
+// queueEndpoints queues the Endpoints resource for processing.
+func (c *Controller) queueEndpoints(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+// onEndpointsAdd queues a sync for the relevant Endpoints resource if
+// skipMirror() is not true.
+func (c *Controller) onEndpointsAdd(obj interface{}) {
+	endpoints := obj.(*v1.Endpoints)
+	if endpoints == nil {
+		utilruntime.HandleError(fmt.Errorf("Invalid Endpoints provided to onEndpointsAdd()"))
+		return
+	}
+	if skipMirror(endpoints.Labels) || forLeaderElection(endpoints.Annotations) {
+		return
+	}
+	c.queueEndpoints(obj)
+}
+
+// onEndpointsUpdate queues a sync for the relevant Endpoints resource if
+// skipMirror() is not true for prev and current resources.
+func (c *Controller) onEndpointsUpdate(prevObj, obj interface{}) {
+	endpoints := obj.(*v1.Endpoints)
+	prevEndpoints := prevObj.(*v1.Endpoints)
+	if endpoints == nil || prevEndpoints == nil {
+		utilruntime.HandleError(fmt.Errorf("Invalid Endpoints provided to onEndpointsUpdate()"))
+		return
+	}
+	if (skipMirror(endpoints.Labels) && skipMirror(prevEndpoints.Labels)) || forLeaderElection(endpoints.Annotations) {
+		return
```

Comment:
I think logging at v(4) could might drown out other more important logs at the same level. Since I'm assuming we'll be ignoring significantly more Endpoints than we're mirroring, often 100%. I've added logging at v(5), but open to feedback on that.